azurerm/resource_arm_container_group.go

```diff
@@ -0,0 +1,323 @@
+		d.Set("ip_address", address.IP)
+	}
+
+	containerConfigs := flattenContainerGroupContainers(resp.Containers)
+	d.Set("container", containerConfigs)
+
+	return nil
+}
+
+func resourceArmContainerGroupDelete(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*ArmClient)
+	containterGroupsClient := client.containerGroupsClient
+
+	id, err := parseAzureResourceID(d.Id())
+
+	if err != nil {
+		return err
+	}
+
+	// container group properties
+	resGroup := id.ResourceGroup
+	name := id.Path["containerGroups"]
+
+	resp, err := containterGroupsClient.Delete(resGroup, name)
+	if err != nil {
+		if utils.ResponseWasNotFound(resp.Response) {
+			return nil
+		}
+		return err
+	}
+
+	return nil
+}
+
+func flattenContainerGroupContainers(containers *[]containerinstance.Container) []interface{} {
+
+	containerConfigs := make([]interface{}, 0, len(*containers))
+	for _, container := range *containers {
+		containerConfig := make(map[string]interface{})
+		containerConfig["name"] = *container.Name
+		containerConfig["image"] = *container.Image
+
+		resourceRequests := *(*container.Resources).Requests
+		containerConfig["cpu"] = *resourceRequests.CPU
+		containerConfig["memory"] = *resourceRequests.MemoryInGB
+
+		if len(*container.Ports) > 0 {
+			containerConfig["port"] = *(*container.Ports)[0].Port
+		}
+		// protocol isn't returned in container config
+
+		containerConfigs = append(containerConfigs, containerConfig)
+	}
+
+	return containerConfigs
+}
+
+func expandContainerGroupContainers(d *schema.ResourceData) (*[]containerinstance.Container, *[]containerinstance.Port) {
+	containersConfig := d.Get("container").([]interface{})
+	containers := make([]containerinstance.Container, 0, len(containersConfig))
+	containerGroupPorts := make([]containerinstance.Port, 0, len(containersConfig))
+
+	for _, containerConfig := range containersConfig {
+		data := containerConfig.(map[string]interface{})
+
+		// required
+		name := data["name"].(string)
+		image := data["image"].(string)
+		cpu := data["cpu"].(float64)
+		memory := data["memory"].(float64)
+
+		container := containerinstance.Container{
+			Name: &name,
+			ContainerProperties: &containerinstance.ContainerProperties{
+				Image: &image,
+				Resources: &containerinstance.ResourceRequirements{
+					Requests: &containerinstance.ResourceRequests{
+						MemoryInGB: &memory,
+						CPU:        &cpu,
+					},
+				},
+			},
+		}
+
+		if v, _ := data["port"]; v != 0 {
+			port := int32(v.(int))
+
+			// container port (port number)
+			containerPort := containerinstance.ContainerPort{
+				Port: &port,
+			}
+
+			container.Ports = &[]containerinstance.ContainerPort{containerPort}
+
+			// container group port (port number + protocol)
+			containerGroupPort := containerinstance.Port{
+				Port: &port,
+			}
+
+			if v, ok := data["protocol"]; ok {
+				protocol := v.(string)
+				containerGroupPort.Protocol = containerinstance.ContainerGroupNetworkProtocol(strings.ToUpper(protocol))
```

Comment:
does this actually need to be `strings.ToUpper`?

Reply:
It's needed as the ContainerGroupNetworkProtocol enum is composed of only all uppercase strings.